hooks: skip post-command hook more often

derrickstolee opened the pull request:
The postCommand.strategy=worktree-change config option allows for skipping the post-command hook when the Git command doesn't change the worktree. However, sometimes this config isn't loaded before the post-command hook is invoked, causing the hook to run in cases where we'd prefer it to not run.

Examples include: `git version` or `git <typo>`.

The tricky bit is that there are several places where we get here and standard config mechanisms can't load due to not having a `gitdir` in the repository struct. We fix this by:

1. Using the "load early config" mechanism also used by `core.hooksPath`.
2. Skipping the lookup for the sentinel file when there isn't a `gitdir` since we couldn't have written one without it.

The change is given in two commits: the first expands the tests with the existing behavior and the second changes the behavior, showing the impact on those tests.

* [X] This change only applies to microsoft/git specifics (post-command hook)

See #736 and #748 for prior art in this space.

derrickstolee commented on t/t0401-post-command-hook.sh:
```diff
@@ -86,4 +86,44 @@ test_expect_success 'with post-index-change config' '
+	echo ran >post-command.out
+	EOF
+
+	GIT_TRACE2_PERF=1 git -C post-typo-no-pre version &&
```
Oops. `GIT_TRACE2_PERF=1` here is a leftover from local debugging. Will fix after letting the PR builds complete.

dscho replied:
How about leaving it in? That could accelerate debugging in case something goes wrong in this test case in a CI build.

mjcheetham commented:
So with this change, if you have _only_ a `post-command` hook, and _no_ `pre-command` hook, then if there's no `gitdir` it will not run (irrespective of the strategy since we haven't any config to read).

I guess this makes sense. It does mean that if you wanted a hook to, say, aid with typos or commands outside of a repo, then you couldn't use a `post-command` hook for that anymore. Is this a reasonable restriction?

derrickstolee commented:
> So with this change, if you have _only_ a `post-command` hook, and _no_ `pre-command` hook, then if there's no `gitdir` it will not run (irrespective of the strategy since we haven't any config to read).

If `r->gitdir` is NULL, then it's that the repository hasn't been loaded yet. This is due to the builtins using custom settings for how it prepares the repo.

But as I write this reply, I realize that _aliases_ or dash-commands may have some special behavior that must be accounted for. I'm investigating now.

> I guess this makes sense. It does mean that if you wanted a hook to, say, aid with typos or commands outside of a repo, then you couldn't use a `post-command` hook for that anymore. Is this a reasonable restriction?

The post-command hook only works within a `$GITDIR` which is somehow set up even when the config hasn't been loaded. It would never run if the Git command is run without a `$GITDIR` variable _or_ its current directory living within a repo on disk. That won't happen in any case.

derrickstolee committed:
t0401: test post-command for alias, version, typo

When the top-level Git process is an alias, it doesn't load much config
and thus the postCommand.strategy config setting is not loaded properly.
This leads to the post-command hook being run more frequently than we want,
including an alias for 'git add' running the hook even when the worktree
did not change.

Similar state is not loaded by 'git version' or 'git typo'.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

dscho reviewed (dismissed):
Looks good to me!

derrickstolee committed:
hooks: better handle config without gitdir

When the post-command hook runs, we could be in several custom states:

1. The command is a regular builtin, but the repo setup is incomplete.
   (Example: "git version", "git rev-parse HEAD")

2. The command is a dashed command, but the top level process uses a space
   in its call. (Example: "git http-request")

3. The command is an alias that goes to a builtin.

4. The command has no arguments or is only helptext.

Each of these cases leads to a place where we previously had not loaded the
postCommand.strategy config and would execute the post-command hook without
looking for a sentinel file.

There are two fixes here:

First, we use early config parsing so we can get config details without
fully setting up the repository. This is how core.hooksPath works in these
situations.

Second, we need to make sure handle_hook_replacement() is called even when
the repo's gitdir is NULL. This requires a small amount of care to say that
a sentinel file cannot exist if the gitdir isn't set, as we would not have
written one without initializing the gitdir.

This gets all of the desired behaviors we want for this setting without
doing anything extreme with how pre- and post-command hooks run otherwise.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee commented:
Thanks for the quick looks, but the alias scenario showed me some interesting complications. I'm pretty happy with the solution now, so please re-review! There are two commits now, where the first is only adding tests to demonstrate untested behavior. The second makes the behavior change and demonstrates how that impacts the new tests.

dscho commented on t/t0401-post-command-hook.sh:
```diff
@@ -78,12 +78,42 @@ test_expect_success 'with post-index-change config' '
+	rm -f post-command.out post-index-change.out &&
+	git config alias.dd add &&
+	echo more stuff >>file &&
+	git dd file &&
```
s/dd/add/?

derrickstolee replied:
It's an alias that I'm testing.

derrickstolee replied:
`git config alias.dd add` in an earlier line.

dscho replied:
D'oh. I totally missed that. Could I request for the alias to be called `alias.aliased_add`? That way I cannot mistake it for a typo in the future.

dscho replied:
Thank you for indulging me!

dscho commented on hook.c:
```diff
@@ -242,9 +261,15 @@ static int handle_hook_replacement(struct repository *r,
+	enabled = strval && !strcasecmp(strval, "worktree-change");
+	free(strval);
+
+	if (!enabled)
```
An alternative would be to pass `&enabled` instead of `&strval`, drop the `strval` variable completely, and perform the comparison in the `store_post_command_strategy()` callback (which should then probably be renamed to `is_worktree_change_strategy()` or some such.

What do you think?

derrickstolee replied:
That certainly simplifies things because of the need to `xstrdup()` in the current setup.

dscho commented on t/t0401-post-command-hook.sh:
```diff
@@ -80,16 +80,12 @@ test_expect_success 'with post-index-change config' '
 	# and this should work through an alias.
-	rm -f post-command.out post-index-change.out &&
 	git config alias.dd add &&
+	rm -f post-command.out post-index-change.out &&
```
This line should probably be added to the correct spot already in the preceding commit.

dscho reviewed (commented):
I really like the structure and presentation of this PR!

emilyyang-ms reviewed (approved):
The 2 commits demonstrated the PR really well, thanks!

wilbaker commented on hook.c:
```diff
@@ -290,7 +313,7 @@ int run_hooks_opt(struct repository *r, const char *hook_name,
 	/* Interject hook behavior depending on strategy. */
-	if (r && r->gitdir &&
+	if (r &&
 	    handle_hook_replacement(r, hook_name, &options->args))
```
Git code style question/comment:

Can the call to `handle_hook_replacement` be moved up on the same line as `if (r &&`?

Or is it git coding convention to keep function calls that are part of conditions on their own line?

derrickstolee replied:
You're right that there's likely room on the line for it right now. I was optimizing for small diffs but should have considered the possibility of room on the line now.

wilbaker reviewed (approved):
Thanks for handling the alias case as well!

I left one small question about formatting, thanks!

dscho reviewed (approved):
❤️
